### ISSUE TITLE
VerticalLayout major problems, infinite invalidation

### ISSUE THREAD

**Issue by teotigraphix** [BowlerHatLLC/feathers#1398]
Hey,

I also have popup dialogs that are not sizing anymore.

I have been beta testing my apps so I wasn't able to test the final until today.

A lot of my layouts are broke and in infinite validation loops.

I will get more details but this is going to be a problem, I reverted back to beta because it's so bad.

I have a PanelScreen with VerticalLayout, 3 child Layout Groups; I need more time to get an firm example but here is my code if it can help you get started;

```
<?xml version="1.0" encoding="utf-8"?>
<screen:AbstractContentPanelScreen xmlns:fx="http://ns.adobe.com/mxml/2009"
                                   xmlns:screen="caustic.ui.screen.*"
                                   xmlns:machine="caustic.guide.ui.component.machine.*"
                                   xmlns:f="library://ns.feathersui.com/mxml"
                                   title="Project Information"
                                   padding="16"
                                   xmlns:component="caustic.ui.component.*">
    <fx:Script>
        <![CDATA[
            import caustic.ui.theme.core.CStyleNames;
        ]]>
    </fx:Script>

    <screen:layout>
        <f:VerticalLayout gap="16"/>
    </screen:layout>

    <component:VGroup percentWidth="100"
                      gap="8">

        <f:Label text="File name"
                 wordWrap="true"
                 styleName="{CStyleNames.H1_TERITARY_LABEL}">
        </f:Label>

        <f:Label id="fileNameLabel"
                 paddingLeft="8"
                 wordWrap="true">
            <f:layoutData>
                <f:VerticalLayoutData percentWidth="100"/>
            </f:layoutData>
        </f:Label>

    </component:VGroup>

    <component:VGroup percentWidth="100"
                      gap="8">

        <f:Label text="File location"
                 wordWrap="true"
                 styleName="{CStyleNames.H1_TERITARY_LABEL}">
        </f:Label>

        <f:Label id="locationLabel"
                 paddingLeft="8"
                 wordWrap="true">
            <f:layoutData>
                <f:VerticalLayoutData percentWidth="100"/>
            </f:layoutData>
        </f:Label>

    </component:VGroup>

    <component:VGroup percentWidth="100"
                      gap="8">

        <f:Label text="File size"
                 wordWrap="true"
                 styleName="{CStyleNames.H1_TERITARY_LABEL}">
        </f:Label>

        <f:Label id="fileSizeLabel"
                 paddingLeft="8"
                 wordWrap="true">
            <f:layoutData>
                <f:VerticalLayoutData percentWidth="100"/>
            </f:layoutData>
        </f:Label>

    </component:VGroup>

    <component:VGroup percentWidth="100"
                      gap="8">

        <f:Label text="File creation Date"
                 wordWrap="true"
                 styleName="{CStyleNames.H1_TERITARY_LABEL}">
        </f:Label>

        <f:Label id="createdLabel"
                 paddingLeft="8"
                 wordWrap="true">
            <f:layoutData>
                <f:VerticalLayoutData percentWidth="100"/>
            </f:layoutData>
        </f:Label>

    </component:VGroup>

    <component:VGroup percentWidth="100"
                      gap="8">

        <f:Label text="File last modification Date"
                 wordWrap="true"
                 styleName="{CStyleNames.H1_TERITARY_LABEL}">
        </f:Label>

        <f:Label id="modifiedLabel"
                 paddingLeft="8"
                 wordWrap="true">
            <f:layoutData>
                <f:VerticalLayoutData percentWidth="100"/>
            </f:layoutData>
        </f:Label>

    </component:VGroup>

</screen:AbstractContentPanelScreen>
```

**Comment by joshtynjala** [comment]
Can you try building the mxml-v3.0.x branch to see if that fixes your issues? Thanks, Mike!

**Comment by teotigraphix** [comment]
So this was the same as I did last time? You merged the branch correct?

**Comment by joshtynjala** [comment]
Yes, I merged into that branch. Use the code to build a new feathers.swc for your SDK.

**Comment by teotigraphix** [comment]
Ok, that didn't help. Here are some screen shots, the first is with the new code from the MXML branch. Notice the FPS is 20 and the memory is going crazy like 1000's of objects being created and then GC'd.

![capture1](https://cloud.githubusercontent.com/assets/226417/16023662/c9d591e8-318b-11e6-97f2-5024622075b7.PNG)

The image below is from Feathers beta 3. As you can see 60fps, perfect layout, BTW this is the layout I posted above.

![capture2](https://cloud.githubusercontent.com/assets/226417/16023704/028ed328-318c-11e6-9463-e54030beec5a.PNG)

**Comment by teotigraphix** [comment]
Dude sorry for the "non code" repro right now, I am really busy and haven't sat down to nail this yet. I am busy fixing my own bugs as well, haha :)

**Comment by joshtynjala** [comment]
Hmmm... I can reproduce in MXML, but my changes worked as expected when I built the same layout in ActionScript. Weird. Well, at least I can reproduce.

**Comment by teotigraphix** [comment]
Ok, so you are saying the compiler is doing something to the generated code that doesn't exactly mirror actionscript?

**Comment by joshtynjala** [comment]
I just tested 100% identical ActionScript code (no MXML) in both branches, and I can still only reproduce this issue in the MXML branch. This is so weird. I wonder if some merge didn't work or something.

**Comment by joshtynjala** [comment]
My mistake. I switched to the mxml-v3.0.x branch, but forgot to pull the latest changes. I can no longer reproduce using the following code:

	<f:PanelScreen autoSizeMode="stage">
		<f:layout>
			<f:VerticalLayout gap="16"/>
		</f:layout>
		
		<f:LayoutGroup>
			<f:layoutData>
				<f:VerticalLayoutData percentWidth="100"/>
			</f:layoutData>
			<f:layout>
				<f:VerticalLayout gap="8"/>
			</f:layout>
			
			<f:Label text="File name"
					 wordWrap="true">
			</f:Label>
			
			<f:Label id="fileNameLabel"
					 paddingLeft="8"
					 wordWrap="true" text="UntitledDocument1">
				<f:layoutData>
					<f:VerticalLayoutData percentWidth="100"/>
				</f:layoutData>
			</f:Label>
			
		</f:LayoutGroup>
		
		<f:LayoutGroup>
			<f:layoutData>
				<f:VerticalLayoutData percentWidth="100"/>
			</f:layoutData>
			<f:layout>
				<f:VerticalLayout gap="8"/>
			</f:layout>
			
			<f:Label text="File location"
					 wordWrap="true">
			</f:Label>
			
			<f:Label id="locationLabel"
					 paddingLeft="8"
					 wordWrap="true" text="/Users/test/documents/UntitledDocument1">
				<f:layoutData>
					<f:VerticalLayoutData percentWidth="100"/>
				</f:layoutData>
			</f:Label>
			
		</f:LayoutGroup>
		
		<f:LayoutGroup>
			 <f:layoutData>
				 <f:VerticalLayoutData percentWidth="100"/>
			 </f:layoutData>
			 <f:layout>
				 <f:VerticalLayout gap="8"/>
			 </f:layout>
			
			<f:Label text="File size"
					 wordWrap="true">
			</f:Label>
			
			<f:Label id="fileSizeLabel"
					 paddingLeft="8"
					 wordWrap="true" text="364.77 Kb">
				<f:layoutData>
					<f:VerticalLayoutData percentWidth="100"/>
				</f:layoutData>
			</f:Label>
			
		</f:LayoutGroup>
		
		<f:LayoutGroup>
			<f:layoutData>
				<f:VerticalLayoutData percentWidth="100"/>
			</f:layoutData>
			<f:layout>
				<f:VerticalLayout gap="8"/>
			</f:layout>
			
			<f:Label text="File creation Date"
					 wordWrap="true">
			</f:Label>
			
			<f:Label id="createdLabel"
					 paddingLeft="8"
					 wordWrap="true" text="Fri Jun 10 11:59:57 GMT-400 2016">
				<f:layoutData>
					<f:VerticalLayoutData percentWidth="100"/>
				</f:layoutData>
			</f:Label>
			
		</f:LayoutGroup>
		
		<f:LayoutGroup>
			<f:layoutData>
				<f:VerticalLayoutData percentWidth="100"/>
			</f:layoutData>
			<f:layout>
				<f:VerticalLayout gap="8"/>
			</f:layout>
			
			<f:Label text="File last modification Date"
					 wordWrap="true">
			</f:Label>
			
			<f:Label id="modifiedLabel"
					 paddingLeft="8"
					 wordWrap="true" text="Fri Jun 10 11:59:57 GMT-400 2016">
				<f:layoutData>
					<f:VerticalLayoutData percentWidth="100"/>
				</f:layoutData>
			</f:Label>
			
		</f:LayoutGroup>
		
	</f:PanelScreen>

**Comment by teotigraphix** [comment]
Take off autoSizeMode="stage" does that change anything? I pulled the latest changes for sure.

**Comment by joshtynjala** [comment]
No, it still worked for me. I also tried setting width and height explicitly.

**Comment by teotigraphix** [comment]
Well I will try again later making sure I have your last change but the only thing I can see that is different is the PanelScreen is in a DrawersApplication, in a ScreenNavigator.

**Comment by teotigraphix** [comment]
If it still messes up, I won't post back until I can get some type of non "my code" example either in MXML of pure AS.

**Comment by teotigraphix** [comment]
![capture3](https://cloud.githubusercontent.com/assets/226417/16056351/26f33c6a-3243-11e6-9609-569346d1a32f.PNG)

I am up to date. So it looks as if there is something more intrinsic to my setup that is causing this invalidation.

I will have to investigate further, I said I wouldn't comment but I just wanted you to know that I was up to date yesterday and there is nothing to pull today, so my source code is not changing.

**Comment by joshtynjala** [comment]
> git pull upstream mxml

The mxml branch does not currently include any of my bug fixes. Please stick to the mxml-v3.0.x branch.

**Comment by teotigraphix** [comment]
Ah my bad, that might explain somethings, but there still was a bug your could reproduce. :)

I forgot that the mxml branch is like master and you don't merge into that until you are sure of the next version.

**Comment by teotigraphix** [comment]
Nope, that my my mistake on the screen shot, I already am in the mxml-v3.0.0 branch, you can see that in the screenshot above. And this..

![capture4](https://cloud.githubusercontent.com/assets/226417/16058115/d37f2b5e-324a-11e6-8d6f-6be7e15848b1.PNG)

**Comment by joshtynjala** [comment]
mxml-v3.0.0 is a tag for version 3.0.0. You want the mxml-v3.0.x branch, with an x at the end. v3.0.x and mxml-v3.0.x is where I'm currently applying fixes for the upcoming 3.0.1 update.

You were on the mxml-v3.0.x in the previous screenshot, but you pulled to the mxml branch. You want to be on that branch, but also pull to the same branch.

**Comment by teotigraphix** [comment]
Ah ok, this look correct?

![capture5](https://cloud.githubusercontent.com/assets/226417/16058241/8290579e-324b-11e6-8dc5-43023f4bfd8f.PNG)

**Comment by joshtynjala** [comment]
It looks like you've pulled the latest changes from the mxml-v3.0.x branch. However, that part that says "((mxml-v3.0.0))" makes me suspect that you aren't on that branch currently. I would recommend checking out the v3.0.x branch first, and then do your pull.

**Comment by teotigraphix** [comment]
Ah one of those days! My git says there is no branch by that name.

![capture6](https://cloud.githubusercontent.com/assets/226417/16058472/b135472a-324c-11e6-8aee-dfb8a14eca41.PNG)

**Comment by teotigraphix** [comment]
The funny thing is I did compile this code and it's fixed. But my popup is still messed up.

![capture7](https://cloud.githubusercontent.com/assets/226417/16058562/163baf2e-324d-11e6-9e11-afc3ef86fecc.PNG)

As you can see though the PanelScreen is laid out and at 60fps

![capture8](https://cloud.githubusercontent.com/assets/226417/16058580/21a0a4f0-324d-11e6-93bd-1d15eb039be9.PNG)

**Comment by joshtynjala** [comment]
Apparently, this is how you checkout a remote branch:

https://gist.github.com/markSci5/5916003

In your case, you probably replace origin with upstream, and test with mxml-v3.0.x.

**Comment by joshtynjala** [comment]
Can you share some code for that pop-up?

**Comment by teotigraphix** [comment]
Ah this component is a bit funky, I was going to get back to it.

Basically it's a LayoutGroup that has a content pane that I add to with <mx:Declaration>s from a sub class.

```
<LayoutGroup>
  <VerticalLayout/>
  <LayoutGroup id="contentGroup">

....
```

This is the part I add in the contentGroup;

```
        <f:LayoutGroup id="contentGroup"
                       minWidth="400"
                       maxWidth="400"
                       minHeight="75">
            <f:layout>
                <f:VerticalLayout gap="25"/>
            </f:layout>

            <f:ProgressBar id="progressBar"
                           minimum="0"
                           maximum="100">
                <f:layoutData>
                    <f:VerticalLayoutData percentWidth="100"/>
                </f:layoutData>
            </f:ProgressBar>

            <f:Label id="statusLabel"
                     wordWrap="true">
                <f:layoutData>
                    <f:VerticalLayoutData percentWidth="100"/>
                </f:layoutData>
            </f:Label>

        </f:LayoutGroup>
```

This class is more complicated then it needs to be but it was working perfectly. I see I am using max values and not setting any explicit values.

**Comment by teotigraphix** [comment]
On a side note, that gist you shared seemed to work and does make sense to me but all the code was up to date, so what I compiled with the correct source.

![capture9](https://cloud.githubusercontent.com/assets/226417/16058973/005dd946-324f-11e6-8d64-13f2177f1811.PNG)

**Comment by joshtynjala** [comment]
I think what you did was start with the mxml-v3.0.0 tag, and you merged in the latest changes from mxml-v3.0.x. I think that put you in a detached HEAD state where you were on the right commit, but not directly on the branch.

**Comment by teotigraphix** [comment]
Hey Josh, don't waste your time thinking about this one right now.

I have done some non kosher things in my Dialog from a long time ago that i need to fix. Let me rewrite this component to my current standards and we can see where we are at.

**Comment by joshtynjala** [comment]
Okay, I'm going to close this issue for now. You may continue to comment on it (and reopen it) if there are still issues with your dialog after you redo it.

**Comment by teotigraphix** [comment]
Yeah all my Dialogs that subclass Dialog are broken with width. It seems I may be doing something wrong, so that I what I am going to investigate right now.

What it seems to be is the measured width is 10 or something, so the backgroundSkin is laid out but the component width is incorrect, not measuring correctly.

**Comment by teotigraphix** [comment]
You can pretty much call this closed, I have checked all main layouts and they work good. The Dialog seemed to be a problem of me using to many percentWidth = 100 and word wrap combinations, so nothing was pinning I guess. (I rewrote my Dialog based on Panel now and just hide/remove the header or footer if I don't need them like the ProgressDialog)

**Comment by teotigraphix** [comment]
Last follow up, I just tested the Remix app with the clip launcher(really complex list layouts) works great!

Can't wait for the .1 update. ;-) Thanks Josh!

**Comment by joshtynjala** [comment]
Thanks, Mike!
